nibabel/parrec.py

```diff
@@ -1008,13 +1018,71 @@ def get_sorted_slice_indices(self):
             array, and (equivalently) the only dimension of
             ``self.image_defs``.
         """
-        slice_nos = self.image_defs['slice number']
-        is_full = vol_is_full(slice_nos, self.general_info['max_slices'])
-        keys = (slice_nos, vol_numbers(slice_nos), np.logical_not(is_full))
-        # Figure out how many we need to remove from the end, and trim them
-        # Based on our sorting, they should always be last
+        if not self.strict_sort:
+            slice_nos = self.image_defs['slice number']
+            is_full = vol_is_full(slice_nos, self.general_info['max_slices'])
+            keys = (slice_nos, vol_numbers(slice_nos), np.logical_not(is_full))
+            sort_order = np.lexsort(keys)
+        else:
+            # Sort based on a larger number of keys.  This is more complicated
+            # but works for .PAR files that get missorted by the above method
+            slice_nos = self.image_defs['slice number']
+            dynamics = self.image_defs['dynamic scan number']
+            phases = self.image_defs['cardiac phase number']
+            echos = self.image_defs['echo number']
+
+            # try adding keys only present in a subset of .PAR files
+            try:
+                # only present in PAR v4.2+
+                asl_labels = self.image_defs['label type']
+                asl_keys = (asl_labels, )
+            except:
+                asl_keys = ()
+            if not self.general_info['diffusion'] == 0:
+                try:
+                    # only present for .PAR v4.1+
+                    bvals = self.image_defs['diffusion b value number']
+                    bvecs = self.image_defs['gradient orientation number']
+                except:
+                    bvals = self.image_defs['diffusion_b_factor']
+                    # use hash to get a single sortable value per direction
+                    bvecs = [hash(tuple(
+                        a)) for a in self.image_defs['diffusion'].tolist()]
+                diffusion_keys = (bvecs, bvals)
+            else:
+                diffusion_keys = ()
+
+            # Define the desired sort order (last key is highest precedence)
+            keys = (slice_nos, echos, phases) + \
+                diffusion_keys + asl_keys + (dynamics, )
+
+            """
+            Data sorting is done in two stages:
+                - run an initial sort using the keys defined above
+                - call vol_is_full to identify potentially missing volumes
+                - call vol_numbers to assign unique volume numbers if for some
+                  reason the keys defined above don't provide a unique sort
+                  order (e.g. this occurs for the Trace volume in DTI)
```

Comment:
After looking into it a bit more closely, I don't think a non-unique order should ever occur (unless we are missing a sort key) and vol_numbers can be removed in the `strict_sort=True` case.  The Trace volume in DTI does actually have a different bval/bvec combination than any of the other volumes, I just didn't read the columns properly before.

One case where I found non-unique sort order was that I was missing a key for `image_type_mr` so when I saved a .PAR with multiple image types (real, imaginary, magnitude, phase), these didn't get sorted properly and there were multiple copies of the same slice per volume, etc unless vol_numbers was used.  However, if the `image_type_mr` key is added then there is a unique sort order again and it is unneeded.